### PR TITLE
Perform JEDI LETKFOI analysis for snow depth

### PR DESCRIPTION
**Pull request by aerorahul**
**Description**

This PR will allow users to run JEDI-based LETKFOI update for snow depth analysis in the global-workflow.
These jobs are run in a single `landanl` task instead of 3 separate tasks.  The workflow is updated.
Some improvements are made to pull out running JEDI executable as well as tarring diag files in `analysis.py` 
The `initialize`, `execute` and `finalize` methods for `LandAnalysis` are filled up.
static methods to create ensemble and adding increments are created, so they can be used elsewhere if desired.

Updates to `pygw` included in this PR:
-  documentation is added to `fsutils.chdir` to make it clear that it is not a replacement for `os.chdir`.
- a filter `to_f90bool` is added to `jinja` to allow type conversion from python bool to fortran bool.
- documentation in `jinja.py` is slightly better in this PR.

This PR depends on:
https://github.com/NOAA-PSL/land-apply_jedi_incr/pull/3
https://github.com/NOAA-EMC/GDASApp/pull/490

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

**How Has This Been Tested?**
Ran a standalone `landanl.sh` job on Hera.
  
**Checklist**

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources]: @jiaruidong2017 @CoryMartin-NOAA Do we really need exclusive node?

File: parm/parm_gdas/land_jedi_fix.yaml
Comment (aerorahul) [inline parm/parm_gdas/land_jedi_fix.yaml:7]: I could add the `orog` data in this yaml file in the future.
This is a TODO.

File: parm/config/gfs/config.landanl
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.landanl]: Is there any circumstance where this would be different than the `$FRAC_GRID` set in `config.base`, other than one uses Fortran boolean (`.true.`/`.false.`) and one is yes/no? If not, we should just use the one in `config.base`.

File: parm/config/gfs/config.landanl
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.landanl]: If this is determined by `$FRACGRID` (per comment), why is it being set independently?

File: ush/python/pygfs/task/analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/analysis.py]: Indentation here does not match the rest of the docblock (needs another space).

File: ush/python/pygfs/task/analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/analysis.py]: I know `DATA` here is in all caps to match the env var, but we don't do this with other python variables to match their shell counterpart, and it violates standard python conventions. Would like to see this changed to `data`, but won't insist on it.

File: ush/python/pygfs/task/analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/analysis.py]: Same as above RE `DATA`. Also, there are no type hints in the method signature, which does not match the other methods.

File: ush/python/pygfs/task/land_analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/land_analysis.py]: I don't think type hints are necessary for `self`, but if you are going to use them, I think they should match the class, so this would be `LandAnalysis`, not `Analysis`.

File: ush/python/pygfs/task/land_analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/land_analysis.py]: Same as above.

File: ush/python/pygfs/task/land_analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/land_analysis.py:404]: There should be some sort of comment here that these are currently hard-coded for a 2-mem ensemble.

File: ush/python/pygfs/task/land_analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/land_analysis.py:385]: For now this works, because we are only using the method for one variable. But if additional variables were ever added, loading the same NetCDF over and over is inefficient. On the other hand, if the the doc-block is correct and this should only be used for snow depth, perhaps the method name should reflect that: `create_snow_depth_ens()`, and then I'd be less worried about this being extending to multiple variables.

File: ush/python/pygfs/task/land_analysis.py
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/land_analysis.py]: So here we are converting from yes/no into a fortran bool. Only reinforces my above question about using the `$FRAC_GRID` in `config.base`, which would also eliminate the need for this conversion.

**Review by WalterKolczynski-NOAA — DISMISSED** [review]

File: ush/python/pygfs/task/land_analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py]: I am glad you noted this.  I agree the instance of the object is not necessary to be type defined; it is implied.
Will remove.

File: ush/python/pygfs/task/analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/analysis.py]: Good point on the use of `DATA`.
I will replace with `workdir` and henceforth the python rendition of `DATA` will be `workdir`.  `data` is too generic a word and not easy to distinguish from say `datadir` or something.  `rundir` would also work.

File: ush/python/pygfs/task/analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/analysis.py]: Will fix.

File: ush/python/pygfs/task/land_analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py]: `config.FRACGRID` has been cast as `True|False` from `YES|NO` when it entered `python`.
Here it is necessary to be cast as a `F90Bool` for parsing into the Fortran namelist.

File: ush/python/pygfs/task/land_analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py:385]: Since the method is part of `land_analysis.py`, it is not necessary to say `create_snow_depth_ens`.
I would wait till there is more information on what other state variables will they perturb and how to create separate methods or turn this into an interface.

File: ush/python/pygfs/task/land_analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py:404]: The doc-block says, a Fake 2 member ensemble.  Isn't that sufficient?

File: parm/config/gfs/config.landanl
Comment (aerorahul) [reply parm/config/gfs/config.landanl]: I don't know.  @jiaruidong2017 and @CoryMartin-NOAA may have a plan for this.

File: ush/python/pygfs/task/land_analysis.py
Comment (WalterKolczynski-NOAA) [reply ush/python/pygfs/task/land_analysis.py:404]: It's not in the doc-block for this method, and either way it is nice to identify which parts need to be replaced later and which parts don't. But I'll leave it up to you.

File: ush/python/pygfs/task/analysis.py
Comment (WalterKolczynski-NOAA) [reply ush/python/pygfs/task/analysis.py]: Yeah, `workdir` or `rundir` is even better. I hate the `DATA` name at least as much as I hate using `CASE` for resolution.

File: parm/config/gfs/config.landanl
Comment (CoryMartin-NOAA) [reply parm/config/gfs/config.landanl]: I'm guessing it will be the same as the model, so `config.base` but we need to double check with @barlage and/or @ClaraDraper-NOAA

Comment (jiaruidong2017) [reply parm/config/gfs/config.landanl]: I agree with @CoryMartin-NOAA because there are many `frac_grid` found from GFS physics model. In the namelist of `apply_incr_nml`, the `frac_grid` is assigned with `$FRACGRID` (defined as yes/no). With the changes to the variable `frac_grid` from yes/no to Fortran boolean made by @aerorahul for the `apply_jedi_incr` code at PR https://github.com/NOAA-PSL/land-apply_jedi_incr/pull/3, we can replace the `$FRACGRID` with `$FRAC_GRID` (defined at `config.base`). We will double check with @barlage and @ClaraDraper-NOAA

File: parm/config/gfs/config.resources
Comment (jiaruidong2017) [reply parm/config/gfs/config.resources]: I have no idea on this. @CoryMartin-NOAA would you please answer this?

Comment (CoryMartin-NOAA) [reply parm/config/gfs/config.resources]: probably not but I don't know how much memory it uses

File: parm/config/gfs/config.landanl
Comment (jiaruidong2017) [reply parm/config/gfs/config.landanl]: The `SNOWDEPTHVAR` (snwdph/snodl) can be defined by `$FRACGRID` (no/yes).

Comment (aerorahul) [reply parm/config/gfs/config.landanl]: updated.

File: ush/python/pygfs/task/land_analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py]: resolved.  But we still need to do Python to F90 boolean transform.

File: parm/config/gfs/config.landanl
Comment (aerorahul) [reply parm/config/gfs/config.landanl]: defining `FRAC_GRID` is done in `config.base`.  This instance is now removed from `config.landanl`.

File: parm/config/gfs/config.resources
Comment (aerorahul) [reply parm/config/gfs/config.resources]: I will remove the `is_exclusive=True`.  If you need more memory, please determine how much rather than reserving the full node for a 6 processor job.

File: ush/python/pygfs/task/land_analysis.py
Comment (jiaruidong2017) [inline ush/python/pygfs/task/land_analysis.py]: If the first `config.FRAC_GRID` is the same as the second `config.FRAC_GRID`, this line is not necessary, right?

**Review by jiaruidong2017 — COMMENTED** [review]
The `landanl` directory needs to be deleted after the job `landanl` is successfully completed.

File: ush/python/pygfs/task/land_analysis.py
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py]: It is not.
The first `config.FRAC_GRID` is `True|False`.  The one being passed to the jinja template is a boolean and needs to be `.true.|.false.`
Alternately, I can write a jinja2 filter `to_F90bool` and update the template to be `{{ FRAC_GRID | to_F90bool }}`

File: ush/python/pygfs/task/land_analysis.py
Comment (jiaruidong2017) [inline ush/python/pygfs/task/land_analysis.py:170]: This copies the observation data in the IODA format based the predefined observation yaml file. 
However, the IMS observation data in the final IODA format (`gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4`) is not available yet, and will be computed in `prepare_IMS`. The name convention of the original IMS data is `ims{YYYY}{DOY}_4km_v1.3.nc`. Therefore, the following errors were caused: 

```
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/fsutils.py", line 83, in cp
    shutil.copyfile(source, target)
  File "/scratch1/NCEPDEV/da/python/opt/core/miniconda3/4.6.14/envs/gdasapp/lib/python3.7/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/scratch1/NCEPDEV/global/Jiarui.Dong/ptmp/landid8888/gdas.20210323/18/obs/gdas.t18z.ims_snow_2021032318.nc4'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/scripts/exglobal_land_analysis.py", line 23, in <module>
    anl.initialize()
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/logger.py", line 264, in wrapper
    retval = func(*args, **kwargs)
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygfs/task/land_analysis.py", line 170, in initialize
    super().initialize()
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygfs/task/analysis.py", line 37, in initialize
    FileHandler(obs_dict).sync()
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/file_utils.py", line 39, in sync
    sync_factory[action](files)
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/file_utils.py", line 59, in _copy_files
    cp(src, dest)
  File "/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/global-workflow.landid8888/ush/python/pygw/src/pygw/fsutils.py", line 85, in cp
    raise OSError(f"unable to copy {source} to {target}")
OSError: unable to copy /scratch1/NCEPDEV/global/Jiarui.Dong/ptmp/landid8888/gdas.20210323/18/obs/gdas.t18z.ims_snow_2021032318.nc4 to /scratch1/NCEPDEV/stmp2/Jiarui.Dong/RUNDIRS/landid8888/landanl.135218/obs/gdas.t18z.ims_snow_2021032318.nc4
```
One solution is to rename the original IMS data name from `ims{YYYY}{DOY}_4km_v1.3.nc` to `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4`. Any other solutions? Thanks.

Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py:170]: I can't reproduce this error.  Can we discuss offline what the issue is?  
I thought the observations are provided in a yaml.

Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py:170]: The observation is provided at the yaml file sorc/gdas.cd/parm/land/obs/config/ims_snow.yaml:

```
  obsdatain:
    engine:
      type: H5File
      obsfile: $(DATA)/obs/$(OPREFIX)ims_snow_{{ current_cycle | to_YMDH }}.nc4
```
In the t00z, t06z, and t12z cycles, `ims_snow.yaml` is not called. Only in the t18z cycle, the ims snow data is required.

Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py:170]: In ush/python/pygfs/task/analysis.py:

```
def get_obs_dict(self) -> Dict[str, Any]:

        # get observers from master dictionary
        observers = obs_list_config['observers']
        copylist = []
        for ob in observers:
            obfile = ob['obs space']['obsdatain']['engine']['obsfile']
            basename = os.path.basename(obfile)
            copylist.append([os.path.join(self.task_config['COM_OBS'], basename), obfile])
        obs_dict = {
            'mkdir': [os.path.join(self.runtime_config['DATA'], 'obs')],
            'copy': copylist
        }
        return obs_dict
```

Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py:170]: > The observation is provided at the yaml file sorc/gdas.cd/parm/land/obs/config/ims_snow.yaml:
> 
> ```
>   obsdatain:
>     engine:
>       type: H5File
>       obsfile: $(DATA)/obs/$(OPREFIX)ims_snow_{{ current_cycle | to_YMDH }}.nc4
> ```
> 
> In the t00z, t06z, and t12z cycles, `ims_snow.yaml` is not called. Only in the t18z cycle, the ims snow data is required.

It seems that the observation file needs to be cycle dependent.

Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py:170]: Or you can link the `ims{YYYY}{DOY}_4km_v1.3.nc` as `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` for all cycles except `18z`.
This will allow you to retain the same observation yaml.

Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py:170]: We already made the observation yaml file as cycle dependent in config.landanl as: 

```
obs_list_name=gdas_land_adpsfc_only.yaml
if [[ "${cyc}" = "18" ]]; then
    obs_list_name=gdas_land_prototype.yaml
fi
```

While:
**gdas_land_adpsfc_only.yaml:**
```
observers:
- !INC ${OBS_YAML_DIR}/adpsfc_snow.yaml
```

**gdas_land_prototype.yaml:**
```
observers:
- !INC ${OBS_YAML_DIR}/adpsfc_snow.yaml
- !INC ${OBS_YAML_DIR}/ims_snow.yaml
```

Therefore, there are no any issues for the t00z, t06z and t12z cycles. The problem is only at the t18z cycle when `ims_snow.yaml` is included. As you suggested, I can link the `ims{YYYY}{DOY}_4km_v1.3.nc` as `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` to avoid this problem, and later the `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` will be replaced by the computed file from prepare_IMS. @CoryMartin-NOAA Do you think it is okay for this approach? Thanks.

Comment (CoryMartin-NOAA) [reply ush/python/pygfs/task/land_analysis.py:170]: What we really need is IODA to handle missing obs... I would rather not link a wrong file here because that is asking for trouble. We can create a fake file if you want, but I think that is more workflow effort than just doing what @jiaruidong2017 mentioned above and use a different list YAML depending on cycle.

Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py:170]: I suspect simply linking `ims{YYYY}{DOY}_4km_v1.3.nc` as `gdas.t{HH}z.ims_snow_{PDY}{HH}.nc4` won't be enough.  IODA is not just looking for the file to be present, but it also expects it to be a IODA file.  `ims{YYYY}{DOY}_4km_v1.3.nc` is not a IODA file and so the program will abort. 

We can proceed with the approach of including the `ims_snow.yaml` only for the 18z cycle.

@CoryMartin-NOAA Is there an issue open in IODA that communicates this requirement for IODA to handle missing IODA files?  If not, I would recommend opening an issue.

Comment (CoryMartin-NOAA) [reply ush/python/pygfs/task/land_analysis.py:170]: Yes https://github.com/JCSDA-internal/ioda/issues/541

Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py:170]: This issue has been fixed with the change at https://github.com/NOAA-EMC/GDASApp/pull/490#issuecomment-1563363699

**Review by jiaruidong2017 — APPROVED** [review]
The snow DA test run succeeded for all four cycles with the fix at https://github.com/NOAA-EMC/GDASApp/pull/490#issuecomment-1563363699

The `landanl` and `preplandobs` directories need to be deleted after the jobs are successfully completed.

File: ush/python/pygfs/task/land_analysis.py
Comment (WalterKolczynski-NOAA) [reply ush/python/pygfs/task/land_analysis.py]: Would prefer a jinja2 filter. Worried about unintended side effects of altering the config value here. For instance, if this line got called twice and started as False, it would go False → '.false.' → '.true.' because the non-null string would be "truthy" on the second pass:

```
>>> a=False
>>> a = ".true." if a else ".false."
>>> print(a)
.false.
>>> a = ".true." if a else ".false."
>>> print(a)
.true.
```

**Review by WalterKolczynski-NOAA — COMMENTED** [review]
I think I'm happy once the FRAC_GRID conversion to fortran namelist bool gets worked out and the GDAS version gets updated after the dependency PRs are merged.

**Comment by aerorahul** [comment]
@jiaruidong2017 I have implemented a change that @WalterKolczynski-NOAA requested regarding the handling of the type conversion from python to Fortran.
I don't expect this to break as I have tested this function in another branch.  
Please feel free to check.
An update to the GDASApp is included in the PR https://github.com/NOAA-EMC/GDASApp/pull/499

**Comment by jiaruidong2017** [comment]
@aerorahul Your recent change for handling the type conversion works well for my tests. 

I noticed one change need to be reverted in `workflow/applications/gfs_cycled.py` as below: 

```
Plase make a change from: 
           configs += ['preplandobs', 'landanlinit', 'landanlrun', 'landanlfinal']
to: 
           configs += ['preplandobs', 'landanl']

And make a change from:
            gdas_gfs_common_tasks_before_fcst += ['preplandobs', 'landanlinit', 'landanlrun', 'landanlfinal']
to: 
            gdas_gfs_common_tasks_before_fcst += ['preplandobs', 'landanl']
```
I think the changes may come from the "merge to develop". Please revert the above changes.

**Comment by aerorahul** [comment]
> @aerorahul Your recent change for handling the type conversion works well for my tests.
> 
> I noticed one change need to be reverted in `workflow/applications/gfs_cycled.py` as below:
> 
> ```
> Plase make a change from: 
>            configs += ['preplandobs', 'landanlinit', 'landanlrun', 'landanlfinal']
> to: 
>            configs += ['preplandobs', 'landanl']
> 
> And make a change from:
>             gdas_gfs_common_tasks_before_fcst += ['preplandobs', 'landanlinit', 'landanlrun', 'landanlfinal']
> to: 
>             gdas_gfs_common_tasks_before_fcst += ['preplandobs', 'landanl']
> ```
> 
> I think the changes may come from the "merge to develop". Please revert the above changes.

Thanks @jiaruidong2017 for catching this.  I have updated this.

**Comment by aerorahul** [comment]
@CoryMartin-NOAA 
gdasApp did not update the hash in feature/stable-nightly following the merge of PR NOAA-EMC/GDASApp#499

**Review by jiaruidong2017 — APPROVED** [review]
This PR is good to me, and hope it can be merged soon. Thanks.

**Comment by aerorahul** [comment]
> This PR is good to me, and hope it can be merged soon. Thanks.

Yes. As soon as the GDASApp hash in `feature/stable-nightly` is updated to include commits from NOAA-EMC/GDASApp#499, I will make that commit and merge.

**Comment by aerorahul** [comment]
And ofcourse a review from @CoryMartin-NOAA

**Comment by CoryMartin-NOAA** [comment]
@aerorahul commit is [1da2e63](https://github.com/NOAA-EMC/GDASApp/commit/1da2e6350609149cec8f21be4585e88494daf9ce)

I will review later this AM

File: Externals.cfg
Comment (CoryMartin-NOAA) [inline Externals.cfg]: update to [1da2e63](https://github.com/NOAA-EMC/GDASApp/commit/1da2e6350609149cec8f21be4585e88494daf9ce)

File: sorc/checkout.sh
Comment (CoryMartin-NOAA) [inline sorc/checkout.sh]: [1da2e63](https://github.com/NOAA-EMC/GDASApp/commit/1da2e6350609149cec8f21be4585e88494daf9ce)

File: ush/python/pygfs/task/analysis.py
Comment (CoryMartin-NOAA) [inline ush/python/pygfs/task/analysis.py:250]: note to myself and @RussTreadon-NOAA we can use this in aero/atm once this PR is merged

File: ush/python/pygfs/task/land_analysis.py
Comment (CoryMartin-NOAA) [inline ush/python/pygfs/task/land_analysis.py:29]: good question, probably not? but let's leave it for now

File: ush/python/pygw/src/pygw/jinja.py
Comment (CoryMartin-NOAA) [inline ush/python/pygw/src/pygw/jinja.py:133]: this is nice!

**Review by CoryMartin-NOAA — COMMENTED** [review]
Everything looks good, thanks @aerorahul , just will approve once GDASApp hash is updated

File: Externals.cfg
Comment (aerorahul) [reply Externals.cfg]: done.

File: sorc/checkout.sh
Comment (aerorahul) [reply sorc/checkout.sh]: done.

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**
